### ISSUE TITLE
Support non-null terminated strings in ddlog's C API

### ISSUE THREAD

**Issue by antoninbas** [vmware/differential-datalog#515]
`ddlog.h` can be used when invoking ddlog from other programming languages. Such languages - such as Go - may not be able to easily / cheaply generate a null-terminated string  from their own string type.

For example Go provides `_GoStringPtr` (https://golang.org/cmd/cgo/) which gives access to the underlying byte array (should be treated as immutable), but that array may not be null-terminated. Which means that in order to call `ddlog_string`, one has to make a copy of the string first (e.g. with `CString`) and ensure it is null terminated. Given that ddlog makes its own copy of the string when building the record, we are up to 2 copies.

It would be great to provide a version of `ddlog_string` (e.g. `ddlog_string_non_null`) which takes the length of the string as an extra parameter and does not assume that the string is null-terminated. The same applies to `ddlog_struct`, even . though in this case one can find a way to use `ddlog_struct_static_cons`.

**Comment by ryzhyk** [comment]
@antoninbas , what do you think about these API changes to support non-null-terminated strings?

```
+/*
+ * Get DDlog table id by name.  The table name is a UTF8 string represented by
+ * its start address and length in bytes.
+ *
+ * On error, returns -1.
+ */
+extern table_id ddlog_get_table_id_with_length(const char* tname, size_t len);
+
+/*
+ * Get DDlog table name from id.
+ *
+ * The `len` argument is used to return the length of the table name in bytes
+ * and must not be `NULL`.
+ *
+ * On success, returns a non-null-terminated UTF8 string.  The length of
+ * the string in bytes is stored in `len`.  On error returns `NULL`.
+ */
+extern const char* ddlog_get_table_name_with_length(table_id id,
+                                                    size_t *len);
+
+/*
+ * Dump record into a non-null-terminated string for debug printing.
+ *
+ * The `len` argument is used to return the length of the string and must not
+ * be `NULL`.
+ *
+ * Returns `NULL` on error.
+ *
+ * The returned string must be deallocated using `ddlog_string_free()`.
+ */
+extern char* ddlog_dump_record_with_length(const ddlog_record *rec,
+                                           size_t *len);
+
+/*
+ * Deallocate a non-null-terminated string returned by DDlog
+ * (currently only applicable to the string returned by
+ * `ddlog_dump_record_with_length()`).
+ */
+extern void ddlog_string_free_with_length(char *s, size_t len);
+
+/*
+ * Create a string value.
+ *
+ * `s` - points to the start of a UTF8 string.  The string does not have to be
+ *       NULL-terminated.
+ * `length` - length of string in bytes.
+ *
+ * This function copies `s` to an internal buffer, so the caller is responsible for
+ * deallocating `s` if it was dynamically allocated.
+ *
+ * Returns `NULL` if `s` is not a valid UTF8 string.
+ */
+extern ddlog_record* ddlog_string_with_length(const char * s, size_t len);
+
-extern const char * ddlog_get_str_non_nul(const ddlog_record *rec);
+extern const char * ddlog_get_str_with_length(const ddlog_record *rec, size_t *len);
+ /*
+ * Same as `ddlog_struct()`, but passes constructor name as
+ * non-null-terminated string represented by its start address and length in
+ * bytes.
+ */
+extern ddlog_record* ddlog_struct_with_length(const char* constructor,
+                                              size_t constructor_len,
+                                              ddlog_record ** args,
+                                              size_t len);
+
+/*
+ * Same as ddlog_struct_static_cons(), but passes constructor name as
+ * non-null-terminated string represented by its start address and length in
+ * bytes.
+ */
+extern ddlog_record* ddlog_struct_static_cons_with_length(
+        const char *constructor,
+        size_t constructor_len,
+        ddlog_record **args, size_t args_len);
+
-extern const char * ddlog_get_constructor_non_null(const ddlog_record *rec,
-                                                   size_t *len);
-
+extern const char * ddlog_get_constructor_with_length(const ddlog_record *rec,
+                                                      size_t *len);
```

**Comment by antoninbas** [comment]
IMO, this is only required for strings which are passed to ddlog. So if you want to keep the API simpler, you don't really need functions like `ddlog_dump_record_with_length`. Really for my use case I only need `ddlog_string_with_length`. `ddlog_get_table_id_with_length` also avoids a copy, but it only really matters if you need to call this function often - which I don't think is the case.

`ddlog_struct_with_length` / `ddlog_struct_static_cons_with_length` can also be useful, even though `ddlog_struct_static_cons` was enough for me. I would recommend *not* adding them, because I think they make the API a bit messy. For my use-case I was planning to pre-allocate one null-terminated string for each constructor, at initialization time. And by the way I can do the same thing for table names.

So basically I'm only advocating for `ddlog_string_with_length`. I don't see a compelling use case for the other ones. Even for APIs where the client is providing strings to ddlog, if they use "static" strings (constructor / table names) one can do without a `with_length` variant, as it is trivial to pre-allocate null-terminated strings and store them as global variables.

**Comment by ryzhyk** [comment]
Part of the motivation for having `_with_length` counterparts for all string functions was programmer convenience for those working in languages other than C.  But I guess this is not particularly relevant, as these people will use wrappers that must only be developed once.

So I will keep `ddlog_string_with_length`. I think I will also keep `ddlog_struct_with_length` / `ddlog_struct_static_cons_with_length`, as this will eliminate the need for pre-allocating null-terminated strings and will thus enable a cleaner Go API.

I will ditch the others.

**Comment by antoninbas** [comment]
Actually when designing a generic (program-agnostic) Go interface like this:
```
func RecordStruct(constructorName string, fields ...Record) Record {
    // ...
}
```
It's pretty hard not to make a copy, as you don't know what's the lifetime of `constructorName` and when the underlying memory will be garbage collected. Odds are it is is defined as a static global variable somewhere, but there is no guarantee. So the Go implementation either has to make a copy and call `ddlog_struct_static_cons` or let ddlog make a copy.

**Comment by ryzhyk** [comment]
I see. Sounds like the Go type system does not have a way to represent statically allocated strings.

Well, at least if you or someone else decides to go down the path of designing such a generic API,  `ddlog_struct_with_length` only requires one copy vs `ddlog_struct`, which involves two (go string -> C string -> Rust string).
